src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -409,13 +442,74 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                     if (_cacheFullDocumentDiagnostics && !span.HasValue)
                     {
                         var state = stateSet.GetOrCreateActiveFileState(_document.Id);
-                        var data = new DocumentAnalysisData(version, diagnostics);
+                        var data = new DocumentAnalysisData(version, _text.Lines.Count, diagnostics);
                         state.Save(executor.AnalysisScope.Kind, data);
                     }
                 }
 
                 if (incrementalAnalysis)
                     _owner._incrementalMemberEditAnalyzer.UpdateDocumentWithCachedDiagnostics((Document)_document);
+
+                async Task<bool> TryDeprioritizeAnalyzerAsync(DiagnosticAnalyzer analyzer, DocumentAnalysisData existingData)
+                {
+                    // PERF: In order to improve lightbulb performance, we perform de-prioritization optimization for certain analyzers
+                    // that moves the analyzer to a lower priority bucket. However, to ensure that de-prioritization happens for very rare cases,
+                    // we only perform this optimizations when following conditions are met:
+                    //  1. We are performing semantic span-based analysis.
+                    //  2. We are processing 'CodeActionRequestPriority.Normal' priority request.
+                    //  3. Analyzer registers certain actions that are known to lead to high performance impact due to its broad analysis scope,
+                    //     such as SymbolStart/End actions and SemanticModel actions.
+                    //  4. Analyzer did not report a diagnostic on the same line in prior document snapshot.
+
+                    // Conditions 1. and 2.
+                    if (kind != AnalysisKind.Semantic ||
+                        !span.HasValue ||
+                        _priorityProvider.Priority != CodeActionRequestPriority.Normal)
+                    {
+                        return false;
+                    }
+
+                    Debug.Assert(span.Value.Length < _text.Length);
+
+                    // Condition 3.
+                    // Check if this is a candidate analyzer that can be de-prioritized into a lower priority bucket based on registered actions.
+                    if (!await CodeActionRequestPriorityProvider.IsCandidateForDeprioritizationBasedOnRegisteredActionsAsync(analyzer, _compilationWithAnalyzers, cancellationToken).ConfigureAwait(false))
+                    {
+                        return false;
+                    }
+
+                    // Condition 4.
```

Comment:
I have verified that the logic added here works fine - when doing edits on the same line with an existing diagnostic from a SymbolStart/SemanticModel analyzer in prior snapshot, we correctly detect the line intersection and return false below. However, for doing operations such as adding newlines or pasting a large piece of code, the below check fails and we can no longer perform a meaningful comparison with prior snapshot to determine if the line where lightbulb is invoked had a diagnostic in prior snapshot.

Reply:
yeah.  it's a pity (esp. as in the editor we can track this stuff forward trivially).  

We could make this work in two ways as a i see it:

1. you import a new optional ITextSpanTrackingService.  This functions in VS by going down to the underlying ITextSNapshots to map the old span forward to the new snapshot.
2. we could attempt to do this *purely* using SourceTexts.  To do this, you would need the old/current source-text, and you'd then GetTextChanges on them to see if you got a reasonable text change back (e.g. not the "the whole thing changed" change).  You'd then adjust the line number accoridngly.